### PR TITLE
Agda cubical

### PR DESCRIPTION
**Pull request by ryanorendorff**
###### Motivation for this change
 
Adds the cubical type theory library to nixpkgs. I asked the maintainers and @Saizan gave approval to add this library to nixpkgs.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests)) (checked for any tests, there do not appear to be).
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`) (all agda files compile)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date (I cannot find a place in the documentation for Agda things)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @jwiegley @laMudri

### CODE REVIEW

**Comment by laMudri** [comment]
Possibly some modifications will be needed when #76653 is merged. /cc @alexarice

**Comment by alexarice** [comment]
This would need to be modified slightly, either this can be merged and I rebase on top of it or the other way round

**Comment by alexarice** [comment]
Might be easier to merge this first

**Comment by ryanorendorff** [comment]
Doesn't matter to me. I can also modify this if the other PR is merged first.

**Comment by veprbl** [comment]
@ryanorendorff This will need to be adapted for the new Agda infrastructure. See manual https://github.com/NixOS/nixpkgs/blob/d30e2468e0c7875d3d4d47404f52647ccea76fcf/doc/languages-frameworks/agda.section.md

**Comment by alexarice** [comment]
Note that v0.1 won't compile with agda 2.6.1 and there is not a more recent version

**Comment by veprbl** [comment]
@alexarice This warrants getting more versions of Agda in nixpkgs?

**Comment by alexarice** [comment]
Perhaps, I feel the libraries should release a version that works with the new agda version, as the master versions of these libraries work

**Comment by ryanorendorff** [comment]
We could choose some development commit that passes their integration server for now.

**Comment by alexarice** [comment]
I feel this is the sensible option

**Comment by ryanorendorff** [comment]
Ok I’ll find a good commit and update the PR when it builds correctly.

**Comment by Saizan** [comment]
I ended up tagging a recent master as the 0.2 release, it does work with 2.6.1:
https://github.com/agda/cubical/releases/tag/v0.2

**Comment by alexarice** [comment]
@Saizan I tried out that commit and get the following error:
```
these derivations will be built:
  /nix/store/1xsk4ln5yi13y5s0cn1i57ccl42nchj9-cubical-0.2.drv
building '/nix/store/1xsk4ln5yi13y5s0cn1i57ccl42nchj9-cubical-0.2.drv'...
unpacking sources
unpacking source archive /nix/store/r5mlcnbhcp77y5wn7xxhggyi1wsaqnas-source
source root is source
patching sources
configuring
no configure script, doing nothing
building
runhaskell ./Everythings.hs gen-except Core Foundations Codata Experiments
Everythings.hs: ./Cubical/Data/DescendingList/Strict.agda: hGetContents: invalid argument (invalid byte sequence)
make: *** [GNUmakefile:30: gen-everythings] Error 1
builder for '/nix/store/1xsk4ln5yi13y5s0cn1i57ccl42nchj9-cubical-0.2.drv' failed with exit code 2
error: build of '/nix/store/1xsk4ln5yi13y5s0cn1i57ccl42nchj9-cubical-0.2.drv' failed
```

**Comment by ryanorendorff** [comment]
I just got the same error. I'll push up what I have at the moment.

**Comment by alexarice** [comment]
The problem doesn't happen outside a nix-build

**Comment by Saizan** [comment]
I believe the script expects the locale to be UTF-8, is that something that can be set for the build?

**Comment by alexarice** [comment]
I think it can

**Comment by Saizan** [comment]
One can test it with `ghc -e "print System.IO.localeEncoding"`.

However `make all` is currently doing testing, it's not something useful for "installation"

File: pkgs/development/libraries/agda/cubical/default.nix [hunk 1]
```diff
@@ -0,0 +1,33 @@
+{ stdenv, mkDerivation, fetchFromGitHub, ghcWithPackages }:
```
Comment (alexarice) [inline pkgs/development/libraries/agda/cubical/default.nix:1]:
```suggestion
{ lib, mkDerivation, fetchFromGitHub, ghcWithPackages }:
```

File: pkgs/development/libraries/agda/cubical/default.nix [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    patchShebangs
+    make
+  '';
+
+  sourceDirectories = [ "." ];
+
+  meta = with stdenv.lib; {
+    description =
+      "A cubical type theory library for use with the Agda compiler";
+    license = stdenv.lib.licenses.mit;
```
Comment (alexarice) [inline pkgs/development/libraries/agda/cubical/default.nix:29]:
```suggestion
    license = licenses.mit;
```

**Comment by ryanorendorff** [comment]
The output from the `localeEncoding` is `ASCII`. I'll try to change it to UTF-8.

File: pkgs/development/libraries/agda/cubical/default.nix [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    platforms = stdenv.lib.platforms.unix;
```
Comment (alexarice) [inline pkgs/development/libraries/agda/cubical/default.nix:30]:
```suggestion
    platforms = platforms.unix;
```

File: pkgs/development/libraries/agda/cubical/default.nix [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    maintainers = with maintainers; [ ryanorendorff ];
```
Comment (alexarice) [inline pkgs/development/libraries/agda/cubical/default.nix:31]: Can you add me as a maintainer

File: pkgs/development/libraries/agda/cubical/default.nix (hunk 3 above)
Comment (alexarice) [inline pkgs/development/libraries/agda/cubical/default.nix:30]: Missing a homepage field. I believe `homepage = src.meta.homepage` should work

**Review by alexarice — CHANGES_REQUESTED** [review]

File: maintainers/maintainer-list.nix [hunk 5]
```diff
@@ -6700,6 +6700,11 @@
     githubId = 889991;
     name = "Ryan Artecona";
   };
+  ryanorendorff = {
+    email = "12442942+ryanorendorff@users.noreply.github.com";
```
[the remark this answers is not in the document]
Comment (ryanorendorff) [reply maintainers/maintainer-list.nix:6704]: Added. Thanks!

@Saizan Do you want to be added as a maintainer?

File: pkgs/development/libraries/agda/cubical/default.nix (hunk 4 above)
Comment (ryanorendorff) [reply pkgs/development/libraries/agda/cubical/default.nix:31]: Sure!

**Comment by alexarice** [comment]
I think `LC_ALL = "en_US.UTF-8";` is what we are looking for with the locale

File: pkgs/development/libraries/agda/cubical/default.nix [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+  # The cubical library has several `Everything.agda` files, which are
+  # compiled through the make file they provide.
+  nativeBuildInputs = [ (ghcWithPackages (self: [ ])) ];
+  buildPhase = ''
+    patchShebangs
+    make
+  '';
+
+  sourceDirectories = [ "." ];
```
[the remark this answers is not in the document]
Comment (ryanorendorff) [reply pkgs/development/libraries/agda/cubical/default.nix:24]: Removed.

File: pkgs/development/libraries/agda/cubical/default.nix [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    owner = "agda";
+    rev = "v${version}";
+    sha256 = "07qlp2f189jvzbn3aqvpqk2zxpkmkxhhkjsn62iq436kxqj3z6c2";
+  };
+
+  # The cubical library has several `Everything.agda` files, which are
+  # compiled through the make file they provide.
+  nativeBuildInputs = [ (ghcWithPackages (self: [ ])) ];
+  buildPhase = ''
+    patchShebangs
```
[the remark this answers is not in the document]
Comment (ryanorendorff) [reply pkgs/development/libraries/agda/cubical/default.nix:20]: No I just added it temporarily while I was matching what the `standard-library` when first making the changes to the new agda setup.

Removed.

File: pkgs/development/libraries/agda/cubical/default.nix [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  src = fetchFromGitHub {
+    repo = pname;
+    owner = "agda";
+    rev = "v${version}";
+    sha256 = "07qlp2f189jvzbn3aqvpqk2zxpkmkxhhkjsn62iq436kxqj3z6c2";
+  };
+
+  # The cubical library has several `Everything.agda` files, which are
+  # compiled through the make file they provide.
+  nativeBuildInputs = [ (ghcWithPackages (self: [ ])) ];
```
[the remark this answers is not in the document]
Comment (ryanorendorff) [reply pkgs/development/libraries/agda/cubical/default.nix:18]: I agree. I had it this way while I was debugging. I'll remove this once the build is working with no packages.

File: maintainers/maintainer-list.nix (hunk 5 above)
[the remark this answers is not in the document]
Comment (Saizan) [reply maintainers/maintainer-list.nix:6704]: I have never used nix myself, so it probably wouldn't help much for me to be a maintainer.

File: pkgs/development/libraries/agda/cubical/default.nix (hunk 3 above)
Comment (ryanorendorff) [reply pkgs/development/libraries/agda/cubical/default.nix]: Added using the suggested statement.

**Comment by ryanorendorff** [comment]
> I think `LC_ALL = "en_US.UTF-8";` is what we are looking for with the locale

Hmm this leads to fun other errors:

```
runhaskell ./Everythings.hs gen-except Core Foundations Codata Experiments
/nix/store/81wybawvkr95c7j8gj5ab3y740mq1fli-bash-4.4-p23/bin/sh: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
/nix/store/81wybawvkr95c7j8gj5ab3y740mq1fli-bash-4.4-p23/bin/sh: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
/nix/store/81wybawvkr95c7j8gj5ab3y740mq1fli-bash-4.4-p23/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
/nix/store/81wybawvkr95c7j8gj5ab3y740mq1fli-bash-4.4-p23/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
/nix/store/81wybawvkr95c7j8gj5ab3y740mq1fli-bash-4.4-p23/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
/nix/store/81wybawvkr95c7j8gj5ab3y740mq1fli-bash-4.4-p23/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
Everythings.hs: ./Cubical/Data/DescendingList/Strict.agda: hGetContents: invalid argument (invalid byte sequence)
make: *** [GNUmakefile:30: gen-everythings] Error 1
builder for '/nix/store/kh89hx2jpkz4kfy3rzwsdvvvpwdjhi37-cubical-0.2.drv' failed with exit code 2
```

**Comment by alexarice** [comment]
Try to add `glibcLocales` to `buildInputs'?

**Comment by ryanorendorff** [comment]
> Try to add `glibcLocales` to `buildInputs`? EDIT: Probably `nativeBuildInputs`

Yes that worked. Building now.

File: maintainers/maintainer-list.nix (hunk 5 above)
[the remark this answers is not in the document]
Comment (ryanorendorff) [reply maintainers/maintainer-list.nix:6704]: Okie dokie. :+1:

File: pkgs/development/libraries/agda/cubical/default.nix (hunk 8 above)
[the remark this answers is not in the document]
Comment (ryanorendorff) [reply pkgs/development/libraries/agda/cubical/default.nix]: Replaced with just `ghc`.

**Comment by ryanorendorff** [comment]
Builds correctly and I was able to type check some cubical code that I had.

**Review by alexarice — APPROVED** [review]
Nix-review passes
Package builds on nixos
Looks ready apart from rebase

**Comment by alexarice** [comment]
@GrahamcOfBorg build agdaPackages.cubical

File: pkgs/development/libraries/agda/cubical/default.nix [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  version = "0.2";
+
+  src = fetchFromGitHub {
+    repo = pname;
+    owner = "agda";
+    rev = "v${version}";
+    sha256 = "07qlp2f189jvzbn3aqvpqk2zxpkmkxhhkjsn62iq436kxqj3z6c2";
+  };
+
+  LC_ALL = "en_US.UTF-8";
```
Comment (laMudri) [inline pkgs/development/libraries/agda/cubical/default.nix:16]: Does this not affect nearly every other Agda package? It feels like it should be the default.

Comment (alexarice) [reply pkgs/development/libraries/agda/cubical/default.nix:16]: Maybe though we haven't run into this problem before. Also this only seems to arise on the haskell building of Everything files and I don't think it is to do with `agda` I'm not sure it is correct to add it to the builder

Comment (laMudri) [reply pkgs/development/libraries/agda/cubical/default.nix:16]: Aah, okay, that explains a bit what's going on. I see stdlib's equivalent Haskell file makes sure everything is done in UTF-8, which is probably the right thing to do upstream.

**Comment by alexarice** [comment]
Thanks @ryanorendorff , @Saizan , and @veprbl for merging 😄

**Pull request merged**